plugins/versioning/src/main/kotlin/versioning/VersioningHandler.kt

```diff
@@ -0,0 +1,98 @@
+package org.jetbrains.dokka.versioning
+
+import com.fasterxml.jackson.annotation.JsonProperty
+import com.fasterxml.jackson.databind.ObjectMapper
+import kotlinx.coroutines.*
+import org.jetbrains.dokka.plugability.DokkaContext
+import org.jetbrains.dokka.plugability.configuration
+import org.jetbrains.dokka.plugability.plugin
+import org.jetbrains.dokka.plugability.query
+import org.jetbrains.dokka.templates.TemplateProcessingStrategy
+import org.jetbrains.dokka.templates.TemplatingPlugin
+import java.io.File
+import java.nio.file.Path
+
+interface VersioningHandler : () -> Unit {
+    fun getVersions(): Map<String, Path>
+    fun currentVersion(): Path?
+}
+
+class DefaultVersioningHandler(val context: DokkaContext) : VersioningHandler {
+
+    private val mapper = ObjectMapper()
+
+    private val processingStrategies: List<TemplateProcessingStrategy> = context.plugin<TemplatingPlugin>().query { templateProcessingStrategy }
+
+    private var versions = mutableMapOf<String, Path>()
+
+    private val configuration = configuration<VersioningPlugin, VersioningConfiguration>(context)
+
+    override fun getVersions() = versions
+
+    override fun currentVersion() = configuration?.let { versionsConfiguration ->
+        versions[versionsConfiguration.currentVersion]
+    }
+
+    override fun invoke() {
+        versions = mutableMapOf()
+        configuration?.let { versionsConfiguration ->
+            val output = context.configuration.outputDir
+            val version = versionsConfiguration.currentVersion
+            addVersionDir(version, output)
+            versionsConfiguration.olderVersions?.let {
+                handlePreviousVersions(it, output)
+            }
+            mapper.writeValue(output.resolve(VERSIONS_FILE), Version(version, versions.keys - version))
+        }
+    }
+
+    private fun addVersionDir(version: String, dir: File) {
+        versions[version] = dir.resolve("index.html").toPath()
+    }
+
+    private fun handlePreviousVersions(olderDokkaPath: File, output: File) {
+        assert(olderDokkaPath.isDirectory) { "Supplied previous version $olderDokkaPath is not a directory!" }
+        val children = olderDokkaPath.list().orEmpty().map { olderDokkaPath.resolve(File(it)) }
+        val oldVersion = children.first { it.name == VERSIONS_FILE }.let { file ->
```

Comment:
firstOrNull + some message if not found since it will be super hard to debug later